src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -1120,10 +1120,22 @@ getSlice index len (Array contents start e _) =
 -- to another array. However, in-place reverse can be useful to take adavantage
 -- of cache locality and when you do not want to allocate additional memory.
 --
--- /Unimplemented/
+-- /Pre-release/
 {-# INLINE reverse #-}
-reverse :: Array a -> m Bool
-reverse = undefined
+reverse :: forall m a. (MonadIO m, Storable a) => Array a -> m Bool
+reverse arr = do
+    let l = 0
+        h = length arr - 1
+    swap l h arr
+
+    where
+
+    swap l h arr0 = do
+        if l < h
+        then do
+            swapIndices arr0 l h
+            swap (l+1) (h-1) arr0
+        else return True
```

Comment:
To implement it more efficiently, we should use two pointers instead of the indices. We can keep incrementing the lower pointer and decrement the higher one. We do not need to do validations of indices and computing the pointer from the index which is being done in swapIndices.